scrapy/downloadermiddlewares/httpproxy.py

```diff
@@ -11,13 +15,35 @@
 from scrapy.utils.python import to_bytes
 
 
+def basic_auth_header(username, password, auth_encoding):
```

Comment:
+1 to extract it, though I think it is better to refactor code to use w3lib's [basic_auth_header](https://github.com/scrapy/w3lib/blob/c1a030582ec30423c40215fcd159bc951c851ed7/w3lib/http.py#L81) function instead.

Reply:
Sure, that's ok.

Reply:
Hi, @kmike 

Now I use the one provided in w3lib, the code becomes less and better. 

But [the test case **test_proxy_auth_encoding**](https://github.com/scrapy/scrapy/blob/1.5/tests/test_downloadermiddleware_httpproxy.py#L110-L113) failed. The failure log:

```
Failure
Traceback (most recent call last):
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/internet/defer.py", line 310, in addCallbacks
    self._runCallbacks()
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/trial/_asynctest.py", line 141, in deferTestMethod
    d = self._run(self._testMethodName, result)
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/trial/_asynctest.py", line 112, in _run
    utils.runWithWarningsSuppressed, self._getSuppress(), method)
--- <exception caught here> ---
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/internet/utils.py", line 201, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/python/compat.py", line 463, in reraise
    raise exception.with_traceback(traceback)
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/internet/utils.py", line 197, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/grammy-jiang/PycharmProjects/scrapy/tests/test_downloadermiddleware_httpproxy.py", line 124, in test_proxy_auth_encoding
    self.assertEqual(req.headers.get('Proxy-Authorization'), b'Basic /HNlcjpwYXNz')
  File "/home/grammy-jiang/.virtualenvs/webcrawler/lib64/python3.6/site-packages/twisted/trial/_synctest.py", line 432, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/usr/lib64/python3.6/unittest/case.py", line 846, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.6/unittest/case.py", line 839, in _baseAssertEqual
    raise self.failureException(msg)
builtins.AssertionError: b'Basic _HNlcjpwYXNz' != b'Basic /HNlcjpwYXNz'
```
The log shows the base64 encoding different at the first character. Then I checked the source code, and found base64 encode method is different:

* the original method is used [base64.b64encode](https://github.com/scrapy/scrapy/blob/1.5/scrapy/downloadermiddlewares/httpproxy.py#L34), which is purely encoding without any escape special character
* the one in w3lib is used [base64.urlsafe_b64encode](https://github.com/scrapy/w3lib/blob/v1.19.0/w3lib/http.py#L99), which [docstring](https://github.com/python/cpython/blob/3.6/Lib/base64.py#L114-L116) said:
**The result is returned as a bytes object.  The alphabet uses '-' instead of '+' and '_' instead of '/'.**

That character escape makes the test case failed. And also I go through [the test suit of w3lib.http](https://github.com/scrapy/w3lib/blob/master/tests/test_http.py) - there are some cases for encoding, but it seems that all of them bypass this failure case perfectly.

Is this a issue about the **w3lib.http.basic_auth_header**?

Reply:
@grammy-jiang I'm not sure, I guess it needs some detective-style work with git (and github) to figure out why a different encoding is used for Authorization and Proxy-Authorization headers, maybe one of them is incorrect.

Reply:
Hi, @kmike 

I have gone through the following documents to make the proxy authentication clear:

* [HTTP authentication - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication)
This page introduces nearly all information about authentication in HTTP protocol, and it states that the headers **Authorization** and **Proxy-Authorization** have several schemes (encode or encrypt ways), the **basic** scheme is one of them ([here](http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml) has a list of authentication schemes, maybe in the future all of them will get support by scrapy). Then I keep gathering more information about the **basic** scheme.

  * [HTTP authentication - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#Basic_authentication_scheme)
Yes, there is a section in the same page to talk about **basic** scheme, and there is one sentence:
    > The "Basic" HTTP authentication scheme is defined in RFC 7617, which transmits credentials as user ID/password pairs, encoded using base64.

  * [RFC 7617 - The 'Basic' HTTP Authentication Scheme](https://tools.ietf.org/html/rfc7617)
This RFC should be the guideline for the **basic** scheme, and actually, it has only one sentence:
    > This document defines the "Basic" Hypertext Transfer Protocol (HTTP) authentication scheme, which transmits credentials as user-id/password pairs, encoded using Base64.

  None of them mention the base64 encode should be escaped for url safe. I think it is because the credentials only exist in headers, not in url.

And also I try to find some third parties' tools to test again:

* [Basic Authentication Header Generator](https://www.blitter.se/utils/basic-authentication-header-generator/)

The result from this online tool proves that the **basic** scheme only use base64 encoding, without special character escaping.

Conclusion: the function basic_auth_header in w3lib should use **base64.b64encode**, not **base64.urlsafe_b64encode**.

Reply:
Hi, @kmike 

I change this basic_auth_header back to the original one. I will use the one in **w3lib.http** when it can pass the original test case in Scrapy.